tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,271 @@
+            header_name,
+            make,
+            override_existing: true,
+        }
+    }
+
+    /// Should the header be overriden if the response already contains it?
+    ///
+    /// Defaults to `true`.
+    pub fn override_existing(mut self, override_existing: bool) -> Self {
+        self.override_existing = override_existing;
+        self
+    }
+}
+
+impl<S, M> fmt::Debug for SetResponseHeader<S, M>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("inner", &self.inner)
+            .field("header_name", &self.header_name)
+            .field("override_existing", &self.override_existing)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<ReqBody, ResBody, S, M> Service<Request<ReqBody>> for SetResponseHeader<S, M>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    M: MakeHeaderValue<ResBody> + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, M>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        ResponseFuture {
+            future: self.inner.call(req),
+            header_name: self.header_name.clone(),
+            make: self.make.clone(),
+            override_existing: self.override_existing,
+        }
+    }
+}
+
+/// Response future for [`SetResponseHeader`].
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, M> {
+    #[pin]
+    future: F,
+    header_name: HeaderName,
+    make: M,
+    override_existing: bool,
+}
+
+impl<F, ResBody, E, M> Future for ResponseFuture<F, M>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    M: MakeHeaderValue<ResBody>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let mut res = ready!(this.future.poll(cx)?);
+
+        if let Some(value) = this.make.make_header_value(&res) {
+            if res.headers().contains_key(&*this.header_name) {
+                if *this.override_existing {
+                    res.headers_mut().insert(this.header_name.clone(), value);
+                }
+            } else {
+                res.headers_mut().insert(this.header_name.clone(), value);
+            }
+        }
+
+        Poll::Ready(Ok(res))
+    }
+}
+
+/// Trait for producing header values from responses.
+///
+/// Used by [`SetResponseHeader`].
+///
+/// You shouldn't normally have to implement this trait since its implemented for closures with the
+/// correct type.
+///
+/// It is also implemented directly for `HeaderValue` so if you just want to add a fixed value you
+/// can suply one directly to [`SetResponseHeaderLayer`].
```

Comment:
in general i think it's good to prefer to avoid second-person ("You ...") in docs. maybe:
```suggestion
/// This trait is implemented for closures with the correct type signature. Typically
/// users will not have to implement this trait for their own types.
///
/// It is also implemented directly for [`HeaderValue`]. When a fixed header value
/// should be added to all responses, it can be  supplied directly to 
/// [`SetResponseHeaderLayer`].
///
/// [`HeaderValue`]: https://docs.rs/http/0.2.3/http/header/struct.HeaderValue.html
```
or something?